Initial PR

JAForbes opened the pull request:
Closes #1

This is a backwards compatible version of @paldepind's brilliant library: [union-type](https://github.com/sanctuary-js/sanctuary-union-type).

I haven't started on the readme yet, but I wanted to open the PR as soon as possible so I can get feedback as early as possible.

I've tried to take care of what I could infer first:
- [X] 100% code coverage
- [X] Makefile lint+test+setup
- [X] LICENSE MIT
- [X] Linting with sanctuary-style
- [X] Backwards compatible with paldepind/union-type
- [X] Added circle CI yml

I'm trying to follow suit with existing projects in the organisation.  
Apologies in advance for any missing pieces.

JAForbes committed:
Initial commit

- [X] 100% code coverage
- [X] Makefile lint+test+setup
- [X] LICENSE MIT
- [X] Linting with sanctuary-style
- [X] Backwards compatible with paldepind/union-type
- [X] Added circle CI yml

I'm trying to follow suit with existing projects in the organization.  Apologies in advance for any missing pieces.

JAForbes committed:
Removed comments from old eslint file

davidchambers commented:
This is very exciting. Thank you, @JAForbes, and thank you @paldepind for creating [union-type](https://github.com/paldepind/union-type)!

davidchambers commented on .gitignore:
```diff
@@ -0,0 +1,5 @@
+jsconfig.json
+.eslintrc
+.vscode
+/coverage/
```
These comments will reveal my obsessive nature:
- The final line should be terminated by a newline character (`\n`). You may like to [configure your editor](https://github.com/editorconfig/editorconfig/wiki/Newline-at-End-of-File-Support) to insert this automatically.
- I like to begin all paths with `/` so it's clear exactly where we expect a file or directory to reside. (There are situations in which this is not practical.)
- I like to end paths to directories with `/`.
- I like to list directories before files, and order the paths in each of these groups alphabetically.
- Only files created by the project should be listed. On macOS, for example, directories may contain **.DS_Store** files. One never wants to commit these, so they should be listed in one's [global .gitignore](https://help.github.com/articles/ignoring-files/#create-a-global-gitignore). **.vscode** falls into this category.
- I believe **jsconfig.json** and **.eslintrc** could be removed from the list.

JAForbes replied:
Great points, I'll update my global .gitignore.

davidchambers replied:
:zap:

davidchambers commented on LICENSE:
```diff
@@ -0,0 +1,25 @@
+The MIT License (MIT)
+
+Copyright (c) 2016 Sanctuary
```
Should we include the following as well?

```
Copyright (c) 2015 Simon Friis Vindum
```

If you just drew inspiration from [union-type](https://github.com/paldepind/union-type) this isn't necessary, but if you used its source code as a starting point I believe Simon has copyright. Fortunately union-type is MIT-licensed. :)

/cc @paldepind

JAForbes replied:
Let's add Copyright for Simon (unless he objects), I didn't reference the source until the final stages but the API design is his.

davidchambers replied:
:zap:

davidchambers commented on LICENSE:
```diff
@@ -0,0 +1,25 @@
+WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+OTHER DEALINGS IN THE SOFTWARE.
+
```
Let's delete this extra line. :)

davidchambers replied:
:zap:

davidchambers commented on Makefile:
```diff
@@ -0,0 +1,32 @@
+.PHONY: test
+test:
+	$(ISTANBUL) cover "test/test.js" -- --recursive
+	$(ISTANBUL) check-coverage --branches 98
```
Let's change the requirement from 98% to 100%. Currently sanctuary-js/sanctuary-set#1 specifies `98` because I integrated Istanbul after the fact and we haven't yet added test cases for the untested paths.

davidchambers replied:
:zap:

davidchambers commented on Makefile:
```diff
@@ -0,0 +1,32 @@
+NPM = npm
+
+SRC = $(.)
+TEST = $(shell find test -name '*.js' | sort)
```
`SRC` and `TEST` aren't referenced in the makefile. Let's remove them.

davidchambers replied:
:zap:

davidchambers commented on Makefile:
```diff
@@ -0,0 +1,32 @@
+
+.PHONY: lint
+lint:
+	$(ESLINT) --env node index.js test/test.js
```
It's not necessary in this case, but I make a habit of including `--` to separate keyword arguments from positional arguments:

``` make
    $(ESLINT) --env node -- index.js test/test.js
```

davidchambers replied:
:zap:

davidchambers committed:
add 'use strict'; pragmas

davidchambers commented on circle.yml:
```diff
@@ -0,0 +1,12 @@
+
+test:
+  override:
+    - make lint test
```
`\n`

davidchambers replied:
:zap:

davidchambers commented on index.js:
```diff
@@ -0,0 +1,318 @@
+
+  if (mapped !== t){
+    return mapped;
+  } else if (t.constructor === Function) {
```
Let's not rely on object identity for the reasons given in sanctuary-js/sanctuary#100. `typeof` should be fine here.

davidchambers replied:
:zap:

davidchambers committed:
use conventional def formatting

davidchambers commented on index.js:
```diff
@@ -0,0 +1,301 @@
+  return staticCase(options, this);
+};
+
+const Setup = function({check, ENV = T.env}){
```
I'm opposed to functions with optional arguments (and records with optional fields). I've seen this pattern used in Haskell:

``` haskell
fooWithOpts :: Opts -> Bar -> Baz
foo :: Bar -> Baz
```

`foo` is defined simply by applying `fooWithOpts` to the default options.

We could follow Sanctuary's lead here. Sanctuary provides [`create`](https://sanctuary.js.org/#create), but the exported module is itself the result of applying `create` to the default options. The only downside of this approach is that the user would need to depend on sanctuary-def in order to disable type checking:

``` javascript
const $ = require('sanctuary-def');
const {create} = require('sanctuary-union-type');

const UnionType = create({check: false, env: $.env});
```

We could, though, export `env` (as Sanctuary does):

``` javascript
const {create, env} = require('sanctuary-union-type');

const UnionType = create({check: false, env: env});
```

Regardless of what we decide, we should rename the `check` field to match the name used by sanctuary-def (and Sanctuary).

JAForbes replied:
I agree with all of this.  One note, the naming for `check` was completely informed by:

https://github.com/paldepind/union-type#disabling-type-checking

But 👍 for having a consistent API within sanctuary libraries.

davidchambers replied:
Ah, yes, I had wondered whether `check` came from union-type. Since we'll be providing a different API, we're free to rename the property for consistency. :)

davidchambers replied:
:zap:

davidchambers commented on index.js:
```diff
@@ -0,0 +1,301 @@
+
+    const Type = T.NullaryType(
+      typeName,
+      a => a && a['@@type'] === typeName
```
Let's make the guard explicit by replacing `a &&` with `a != null &&`.

Let's use `x` as the parameter name to avoid shadowing our type variable named `a`.

davidchambers replied:
:zap:

davidchambers commented on index.js:
```diff
@@ -0,0 +1,301 @@
+      env: ENV,
+    });
+
+  const CreateUnionType = function(typeName, rawCases, prototype = {}){
```
Let's avoid this optional parameter. We can ensure that `prototype` is always provided by making this change:

``` diff
 const Named =
   def('UnionType.Named',
       {},
       [T.String, T.StrMap(T.Any), T.Any],
-      CreateUnionType);
+      (x, y) => CreateUnionType(x, y, {}));
```

`x` and `y` could be renamed, of course. :)

davidchambers replied:
:zap:

davidchambers commented on index.js:
```diff
@@ -0,0 +1,301 @@
+      Object.keys(rawCases);
+
+    const env =
+      ENV.concat(Type);
```
I'd prefer `ENV.concat([Type])`.

davidchambers replied:
:zap:

davidchambers commented on package.json:
```diff
@@ -0,0 +1,30 @@
+  "main": "index.js",
+  "directories": {
+    "test": "test"
+  },
```
Is this actually used in any way? The [documentation](https://docs.npmjs.com/files/package.json#directories) is not particularly enlightening.

JAForbes replied:
No idea, I think `npm init` added that automatically.

davidchambers replied:
Okay. Let's remove it in that case.

davidchambers replied:
:zap:

davidchambers committed:
sanctuary-style@0.3.x

davidchambers committed:
package.json: use 1.2.x format for dependency versions

davidchambers commented on test/test.js:
```diff
@@ -0,0 +1,406 @@
+    Point.Point(4, 'foo');
+  } catch (e) {
+    t.equal(err, e.toString());
+  }
```
I'm not sure about this approach. Won't the assertion only be performed if evaluating the `try` block results in an exception? I think we need [`assert.throws`](https://nodejs.org/api/assert.html#assert_assert_throws_block_error_message) here.

We could define a function such as this:

``` javascript
//    throws :: (Function, TypeRep a, String) -> Undefined !
const throws = (f, type, message) => {
  assert.throws(f, err => err.constructor === type && err.message === message);
};
```

We'd use it like so:

``` javascript
throws(() => { Point.Point(4, 'foo'); },
       TypeError,
       '...');
```

davidchambers replied:
:zap:

davidchambers commented on test/test.js:
```diff
@@ -0,0 +1,406 @@
+      100
+      , area(rect)
+    );
+  }
```
Very nice use of block scope! I've never seen this before! :open_mouth:

davidchambers commented on test/test.js:
```diff
@@ -0,0 +1,406 @@
+  /* eslint-disable no-var */
+  var List =
+    Type({Nil: [], Cons: [T.Any, List]});
+  /* eslint-enable no-var */
```
This is confusing. This would be much clearer:

``` javascript
const List = Type({Nil: [], Cons: [T.Any, undefined]});
```

I think we can do better though. How about supporting one of the following?

``` javascript
const List = Type({Nil: [], Cons: [T.Any, UnionType.Self]});
```

``` javascript
const List = Type({Nil: [], Cons: [T.Any, UnionType.Recur]});
```

``` javascript
const List = Type({Nil: [], Cons: [T.Any, UnionType.Recursive]});
```

The value of `UnionType.Self` (or whatever we were to call it) _could_ simply be `undefined`, but I'd prefer something like `{'@@functional/recursive-type-reference': true}`.

davidchambers replied:
Let's discuss this in #4.

JAForbes replied:
I agree its confusing, using `undefined` could create recursive types by accident, but I would really like this library to be a superset of paldepind/union-type's API.

I'm torn, on the one hand removing the `undefined` feature seems like the right API design choice, but then we are no longer supporting the existing API and discussions of API design are suddenly warranted.  That isn't a path I want to go down.

But on the other hand, if there was ever a juncture to make a breaking change it would be now.  

But ultimately I feel there is more value in continuing the trajectory of keeping this library as a super-set of the original.  So I'd prefer to continue to support `undefined` and in addition support `UnionType.self = {'@@functional/recursive-type-reference': true}`.

Its a similar decision to supporting paldepind's anonymous union type's in addition to named union-types.

davidchambers replied:
If we continue to use `undefined` as the sentinel value users are free to alias it:

``` javascript
const Self = undefined;

// ...

const List = Type({Nil: [], Cons: [a, Self]});
```

I could live with this. :)

davidchambers committed:
license: remove blank line at end of file

davidchambers committed:
makefile: remove unused variable declarations

davidchambers committed:
use Unix line endings

davidchambers committed:
makefile: use "--" separator

davidchambers committed:
makefile: increase branch coverage requirement to 100%

davidchambers committed:
package.json: remove "directories"

davidchambers committed:
lint: remove ESLint exemptions

davidchambers committed:
style: use ternary expressions

davidchambers committed:
avoid "inconsistent" use of Array#concat

davidchambers committed:
use explicit "!= null" guard

davidchambers committed:
do not rely on constructor identity

davidchambers committed:
group and alphabetize imports

davidchambers committed:
style: use more compact formatting

davidchambers committed:
style: use conventional identifier for sanctuary-def module

davidchambers committed:
test: replace tape with mocha

davidchambers committed:
test: use assert.throws

davidchambers committed:
update .gitignore

davidchambers committed:
test: rename file

davidchambers committed:
:%s/\<check\>/checkTypes/g

davidchambers committed:
require "env" field

davidchambers committed:
remove boundStaticCase

davidchambers committed:
internal simplifications

davidchambers committed:
more internal simplifications

davidchambers committed:
remove one use of curryN

davidchambers committed:
avoid type coercion

davidchambers committed:
use Z.map

davidchambers committed:
license: add @paldepind as a copyright holder

davidchambers committed:
:%s/\<fn\>/f/g

davidchambers committed:
remove CaseRecordType

davidchambers committed:
update package.json

davidchambers committed:
add CONTRIBUTING.md

davidchambers committed:
makefile: integrate transcribe and xyz

davidchambers committed:
remove BuiltInType

sanctuary-def provides a way to define arbitrary types, so we needn't
support constructors, predicates, and types.

davidchambers committed:
remove special case for typeof x === 'function'

davidchambers committed:
change type of List.Nil from "() -> List a" to "List a"

This requires removing caseOn, which we may decide is not a good idea.
If so, we can easily revert this commit.

davidchambers committed:
style: remove unnecessary { ... } block

davidchambers committed:
test: tweak Maybe examples

davidchambers committed:
strip namespaces

davidchambers committed:
progress towards unification with $.RecordType

davidchambers committed:
remove duplication

davidchambers committed:
test: use type variables

davidchambers committed:
test: add namespace

davidchambers committed:
revert change to handling of nullary data constructors

davidchambers committed:
reinstate support for defining methods by mutating prototype

safareli commented on index.js:
```diff
@@ -4,6 +4,9 @@ const $ = require('sanctuary-def');
 
 
+//    stripNamespace :: String -> String
+const stripNamespace = s => s.slice(s.indexOf('/') + 1);
```
What should happen if input is `my-package/sub-package/Point` ? maybe we should use `lastIndexOf`?

davidchambers replied:
Good question. I've been assuming type identifiers contain one slash (`/`) at most.

safareli commented on test/index.js:
```diff
@@ -176,7 +178,7 @@ test('Destructuring assignment to extract values', () => {
 
 test('Recursive Union Types', () => {
-  const List = Type({Nil: [], Cons: [$.Any, undefined]});
+  const List = Type({Nil: [], Cons: [a, undefined]});
```
I think the recursion could be expressed like this:

``` js
const List = Type({Nil: [], Cons: (List) => [a, List(a)]});
const Tree = Type({Tip: [a], Node: (Tree) => [Tree(a), Tree(a)]});
```

davidchambers commented:
@JAForbes, shall we close this pull request now that [JAForbes/sum-type][1] exists? I will open an initial pull request for a more Sanctuary-style approach when I complete the dozen or so Sanctuary-related tasks currently on my to-do list.


[1]: https://github.com/JAForbes/sum-type

JAForbes commented:
Yes, close away 😊 

-----Original Message-----
From: "David Chambers" notifications@github.com
Sent: ‎11/‎21/‎2016 2:02 PM
To: "sanctuary-js/sanctuary-union-type" sanctuary-union-type@noreply.github.com
Cc: "James Forbes" james.a.forbes@gmail.com; "Mention" mention@noreply.github.com
Subject: Re: [sanctuary-js/sanctuary-union-type] Initial PR (#3)

@JAForbes, shall we close this pull request now that JAForbes/sum-type exists? I will open an initial pull request for a more Sanctuary-style approach when I complete the dozen or so Sanctuary-related tasks currently on my to-do list.
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.
